### PR TITLE
[Impeller] Encode directly to command buffer for Metal.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Part of https://github.com/flutter/flutter/issues/140804

Rather than using impeller::Command, the impeller::RenderPass records most state directly into the Vulkan command buffer. This should remove allocation/free overhead of the intermediary structures and make further improvements to the backend even easier.

This completely removes the background worker threads used for encoding, which should lower overall CPU usage without decreasing performance by much (though it may be a tad slower or a tad faster).

### CODE REVIEW

File: impeller/renderer/backend/metal/render_pass_mtl.h [hunk 1]
```diff
@@ -22,14 +197,31 @@ class RenderPassMTL final : public RenderPass {
   friend class CommandBufferMTL;
 
   id<MTLCommandBuffer> buffer_ = nil;
+  id<MTLRenderCommandEncoder> encoder_ = nil;
   MTLRenderPassDescriptor* desc_ = nil;
   std::string label_;
+  bool is_metal_trace_active_ = false;
```
Comment (jonahwilliams) [inline impeller/renderer/backend/metal/render_pass_mtl.h:203]: TODO: this

Comment (jonahwilliams) [reply impeller/renderer/backend/metal/render_pass_mtl.h:86]: Done

**Review by bdero — APPROVED** [review]
LGTM

File: impeller/renderer/backend/metal/context_mtl.mm [hunk 2]
```diff
@@ -88,24 +88,6 @@ static bool DeviceSupportsComputeSubgroups(id<MTLDevice> device) {
...
   is_gpu_disabled_sync_switch_->AddObserver(sync_switch_observer_.get());
 
-  // Worker task runner.
-  {
-    raster_message_loop_ = fml::ConcurrentMessageLoop::Create(
-        std::min(4u, std::thread::hardware_concurrency()));
-    raster_message_loop_->PostTaskToAllWorkers([]() {
-      // See https://github.com/flutter/flutter/issues/65752
-      // Intentionally opt out of QoS for raster task workloads.
-      [[NSThread currentThread] setThreadPriority:1.0];
```
Comment (chinmaygarde) [inline impeller/renderer/backend/metal/context_mtl.mm:98]: Get rid of the corresponding Foundation.h header import above?

File: impeller/renderer/backend/metal/render_pass_mtl.h [hunk 3]
```diff
@@ -13,6 +13,64 @@
...
+//-----------------------------------------------------------------------------
+/// @brief      Ensures that bindings on the pass are not redundantly set or
+///             updated. Avoids making the driver do additional checks and makes
+///             the frame insights during profiling and instrumentation not
+///             complain about the same.
+///
+///             There should be no change to rendering if this caching was
+///             absent.
+///
+struct PassBindingsCache {
```
Comment (chinmaygarde) [inline impeller/renderer/backend/metal/render_pass_mtl.h:25]: Move to its own TU and rename as `PassBindingsCacheMTL`.

**Review by chinmaygarde — APPROVED** [review]

File: impeller/renderer/backend/metal/render_pass_mtl.h [hunk 4]
```diff
@@ -22,14 +80,31 @@ class RenderPassMTL final : public RenderPass {
...
+  PassBindingsCache pass_bindings_;
+
+  // Per-command state
+  size_t instance_count_ = 1u;
+  size_t base_vertex_ = 0u;
+  size_t vertex_count_ = 0u;
+  bool has_valid_pipeline_ = false;
+  bool has_label_ = false;
+  BufferView index_buffer_;
+  PrimitiveType primitive_type_;
```
Comment (chinmaygarde) [inline impeller/renderer/backend/metal/render_pass_mtl.h:98]: Default this ivars.

Comment (jonahwilliams) [reply impeller/renderer/backend/metal/render_pass_mtl.h]: Done

File: impeller/renderer/backend/metal/render_pass_mtl.h (hunk 3 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/metal/render_pass_mtl.h]: Done

File: impeller/renderer/backend/metal/context_mtl.mm (hunk 2 above)
Comment (jonahwilliams) [reply impeller/renderer/backend/metal/context_mtl.mm:98]: Done

**Pull request merged**